### PR TITLE
Feat/customizable listener

### PR DESCRIPTION
**Pull request by breautek**
closes #8
closes #7 

This PR introduces several breaking changes that I believe will help make the plugin more scalable into future changes and allowing for deeper customisation based application needs.

The breaking changes includes:

### Symbol renames

Fixing this is more of a pet peeve. Class names shouldn't be using plural form and this repo used several plural symbol names. For JS users this impact will be minimal as usage of `window.totalpave.Insets` needs to be refactored anyway. For TS users, the impact might be a bit larger if they explicitly use the interface type names (as they should). The aim is to make a cleaner API following common naming conventions moving forward.

### Inset API

The inset API is no longer static and the plugin does not setup a provider by default anymore. Instead it is the application responsibility to create and maintain a reference to an `Inset` instance. The inset instance can be created using a configuration object which now accepts a boolean to decide if it should include rounded corners in the inset information, as well as a mask value giving the application the flexibility of choosing whatever mask that makes sense for their use case.

### CODE REVIEW

File: CHANGELOG.md
Comment (TylerBreau) [inline CHANGELOG.md]: typo + super short line

**Review by TylerBreau — CHANGES_REQUESTED** [review]
Document the justification and intent behind the changes.
This documentation does not necessarily need to be in the repo itself.

**Comment by breautek** [comment]
> Document the justification and intent behind the changes. This documentation does not necessarily need to be in the repo itself.

Done, was in the process of doing it. I've written a PR description. The main intent is for @seamlink-aalves desire to include the `IME` mask but for our applications that will probably be too much of a breaking change (it will cause an inset update to include the virtual keyboard).

I think the `CHANGELOG` notes for the breaking changes is sufficient for how to migrate from 0.2.x to 0.3.x. When this PR is finalized, I'll be making a PR for our tp-mobile-common repo that does this migration.

It is a bit of a restructure of the plugin, but I think necessary to make this plugin flexible. I didn't want to just add a bunch of booleans for different mask settings and the "old" way made it very easy to introduce race conditions if the application wanted to change the inset settings. So this PR I believe addresses all of those concerns.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: All public APIs should have their appropriate typings here.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Replace $isUpdateEvent and $isInitEvent with enum or constants.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts:24]: Replace $isUpdateEvent with enum or constants.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Move this to its own file.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Honestly, this should also be in its own file because it's a public-facing type.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts:224]: Replace $isInitEvent and $isUpdateEvent with enum or constants.

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Why don't we just have this in the non-static function?

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Why is this public?

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: plural naming

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: plural naming

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: plural naming

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: plural naming

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: Why is this public?

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts]: plural naming (see all references to this private member)

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts:81]: Why don't we just put the static function's code here? What's the benefit to having both APIs?

File: src/www/Inset.ts
Comment (TylerBreau) [inline src/www/Inset.ts:31]: Move id to the base IInsetEvent.
I don't see why any event would be missing the id attribute, including the init event.
data should be a nullable attribute (null for init). Data should remain as a generic type.

init event will not use data and will instead use the id attribute.
IInsetUpdateEvent type the data to IInset and use data.

**Review by TylerBreau — CHANGES_REQUESTED** [review]
java file still uses the plural name

Have not reviewed code, and have not extensively reviewed documentation files. Will do that on the next review. For now there is many javascript changes.
Be sure to update documentation and java as appropriate for the javascript changes.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts]: These methods are typescript with a special signature so that it can change the type of `e` if it returns true.

That's why the method has a `e is X` return type. That way TypeScript will treat `e` as `X` while inside the if condition block.

It cannot be replaced with a constant or enum.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts:24]: Again, these are type definitions. I cannot replace them with runtime values.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts:81]: Mostly convenience.

The static method contains the implementation and you could call `Inset.free(myInset)` or simply call `myInset.free()`

This is akin to stream/file descriptor's `close()` or whatever.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts]: I think TS will allow it to be private since it's coming from the same file even if it's a static method but it's more or less schematics.

I think I'll change it to a private method however.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts:31]: the id is passed back as `data` in the init event. The entire purpose of this event is send back an native identifier back to the webview so it can finish constructing the `Inset` instance.

the `id` field on the update event is there because at that time it's expected that the `Inset` object is already established and can be retrieved by the id. Whereas during the init event it's in the process of constructing that inset instance.

So I don't really agree with this suggestion.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts]: Also kept (but deprecated) `getInsets` in favour of `getInset()`.

File: src/www/Inset.ts
Comment (breautek) [reply src/www/Inset.ts]: I don't really see much point in maintaining this interface so I removed it. The global typedefs can reference the `Inset` class directly.

File: src/www/Inset.ts
Comment (TylerBreau) [reply src/www/Inset.ts:31]: Summarizing outside discussion: The main benefit is that we can very likely guarantee all events will have an id and IMO it's really not that odd for the data attribute to be null for an event or two. At the end of the day though, this is private implementation and we really don't expect to have any new events. So this is low priority regardless of the final decision. It is also easy to change because the event type is private implementation details.

File: src/www/IInsetConfiguration.ts
Comment (TylerBreau) [inline src/www/IInsetConfiguration.ts:23]: I think this should be typed to the InsetMark enum?

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java:35]: I'm assuming you've done your due diligence in ensuring there's no unused imports.

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java]: plural naming

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java]: improper tabbing

**Comment by seamlink-aalves** [comment]
I've not been very helpful on this one. I've tried testing the PR to check the changes but keep getting errors when loading the plugin. 

I think config.xml still need to be adapted to load the new script files. I tried to change it but had no success.

**Comment by TylerBreau** [comment]
> I've not been very helpful on this one. I've tried testing the PR to check the changes but keep getting errors when loading the plugin.
> 
> I think config.xml still need to be adapted to load the new script files. I tried to change it but had no success.

Yea, that's a good call. Doing a thorough review atm, haven't gotten to the docs yet but a quick peek seems like it is missing declarations for some of the new files.

**Comment by breautek** [comment]
> I've not been very helpful on this one. I've tried testing the PR to check the changes but keep getting errors when loading the plugin.
> 
> I think config.xml still need to be adapted to load the new script files. I tried to change it but had no success.

It's been a WIP, I'm still testing stuff on my end here. The `plugin.xml` is generated on `prepack`, im not sure if that's ran automatically on git installs. It may need to be git cloned and then run `npm run build && npm pack` manually and install the built tgz file instead.

Don't forget that cordova doesn't like "upgrades", so the plugin needs to be uninstalled first, then reinstalled.

Assuming you manually built and packed:

```
cordova plugin remove @totalpave/cordova-plugin-insets
cordova plugin add file:../cordova-plugin-insets/totalpave-cordova-plugin-insets-x.y.z.tgz
```

The current commit on this PR **should** work/install but I'm still refactoring stuff in our codebase and testing things.

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java:281]: Concerned about race condition. Imaging possibility that the request apply insets runs, triggers, runs the update code path, before the init code path goes to js land across the cordova bridge.

Also this code path does not guarentee the insets are set within the asynchronous... Train.
It's disconnected, waiting for the update callback.
The update callback should always come eventually, because of the requestApplyInsets but it is breaking out from the promise chains in the javascript land.

So in the current state, initializing the plugin via the create APIs does not guarantee available insets. So it's not finished initializing.

Comment (TylerBreau) [reply src/android/com/totalpave/cordova/inset/Inset.java:281]: What I think should happen to guarentee insets on listener creation is...

1-Remove init and update event. There is only 1 event, InsetEvent, it has id and data attribute. data is always insets data.
2-Remove the "return plugin result" code from the create API. The create API will force an insets update.
3-The native insets update code will work as it has been working so far.
4-The javascript update code path will check to see if the id is setup in javascript land.
If not setup it up.
5-Regardless of what happens in point 4, proceed with the update code path.

Comment (TylerBreau) [reply src/android/com/totalpave/cordova/inset/Inset.java:281]: Leaving the previous comment there as it was an option. But we've decided on a different solution outside of github.

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java]: HashMap.remove(Object Key) returns the object that was removed or null if not present. https://developer.android.com/reference/java/util/HashMap#remove(java.lang.Object)

File: plugin.xml.ejs
Comment (TylerBreau) [inline plugin.xml.ejs]: I guess the repo was renamed at some point and this wasn't changed... lol. I suppose it should be fixed in another update.

File: CHANGELOG.md
Comment (TylerBreau) [inline CHANGELOG.md:35]: getInsets was deprecated in favour of getInset. Please document.

File: README.md
Comment (TylerBreau) [inline README.md:17]: Not sure when the platform guard was added but, the readme is wrong.

iOS results in what is almost a no-op. You can use apis without platform guards with a single cavet - iOS' inset object will always contain 0 for the 4 sides. It is expected that you will use const and env css variables.

This behaviour is almost identical to current master. In current master, you can add listeners on iOS but they will never be called. The only time you had to think about platform guards was if you used the `getInsets` API. Noting this API was on the plugin object, not an Inset class.

This was choice was intentional so that we could explicitly avoid using platform guards in our apps.

This choice was for convenience so if it's problematic or in the way we could get rid of this idea, though I suppose that would be a breaking change as it could break the iOS side if people didn't use platform guards before...

Whether it's in this PR or another PR, we should decide the expectations of the plugin so that everyone is on the same page.

File: docs.md
Comment (TylerBreau) [inline docs.md]: Missing docs on `getInset`

Comment (TylerBreau) [reply docs.md]: Also `getId`

Comment (TylerBreau) [reply docs.md]: Doesn't talk about static free vs not-static free APIs.

File: src/www/IInsetConfiguration.ts
Comment (TylerBreau) [reply src/www/IInsetConfiguration.ts:23]: update docs if so

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java:52]: Is this any different from simple assignment? I mean if you wanted to add another option in the middle, you'd need to update everything below it no?

File: docs.md
Comment (TylerBreau) [inline docs.md]: Insetes typo

File: docs.md
Comment (TylerBreau) [inline docs.md]: was deprecated

**Review by TylerBreau — CHANGES_REQUESTED** [review]
In regards, to all requests to replace the type hardcoded values with enums and the related $is functions... As there is several comments, I will report the final decision on them all here. 
Reviewing the requests, ultimately we either need the functions for the type guard (+inherent type assertion) or to manually type assert. IMO, either way is not particularly great but I also don't have a better solution. All comments for that can be ignored now.

While an enum will work in place of hardcoding init/update strings, I will defer to the general rule of thumb of "2, copy paste, 3+ abstract". These comments can also be ignored.

File: src/www/IInsetConfiguration.ts
Comment (breautek) [reply src/www/IInsetConfiguration.ts:23]: Nah it's a mask.

so if had an enum like:

{
ONE = 1,
TWO = 2
}

and you type it as the enum, it allows for only `1` or `2`.

But `ONE | TWO` is also a valid value, which is a value of 3. That's why it's typed as a number.

File: README.md
Comment (breautek) [reply README.md:17]: imo best to keep it documented as if it's unsafe to call on the API on iOS. I know it is guarded but it still isn't really right.

**Comment by TylerBreau** [comment]
> I've not been very helpful on this one. I've tried testing the PR to check the changes but keep getting errors when loading the plugin.
> 
> I think config.xml still need to be adapted to load the new script files. I tried to change it but had no success.

After reviewing the plugin xml file, I realized that it's referencing a build artifact. The typescript files should be getting compiled into www/insets.js (not to be confused with src/www directory).

ATM, I'm assuming the built file is packaged into the npm release, which is why it's not in the github repo.

File: docs.md
Comment (breautek) [reply docs.md]: I intentionally omitted `getID` since it's not really valuable to end users.

I'll turn rename getInsets, we don't need to document the depecated method.


The create/free APIs are noted as static APIs.

File: src/www/IInsetConfiguration.ts
Comment (TylerBreau) [reply src/www/IInsetConfiguration.ts:23]: Hmm, what's the value of this style of mask compared to a bitmask?

`
enum InsetMask {
   ONE: 1,
   TWO: 2,
   THREE: 4
}
`

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (breautek) [reply src/android/com/totalpave/cordova/inset/Inset.java:52]: There is absolutely no reason to add a line in the middle. These values are replicated and are exclusive to the plugin. They are mapped to the android's sdk constants. This decouples our code from the android's sdk, which doesn't document any explicit constant value for their inset types.

File: docs.md
Comment (breautek) [reply docs.md]: Also corrected `IInsets` symbol in the docs (depluralized)

File: src/www/IInsetConfiguration.ts
Comment (TylerBreau) [reply src/www/IInsetConfiguration.ts:23]: Ah, I guess that way doesn't really allow it be typed. I also didn't realize >> was a bitmask operator, I'm assuming it's another way to setup a bitmask.

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (breautek) [reply src/android/com/totalpave/cordova/inset/Inset.java:35]: Double checked in android studio. everything currently is being used.

**Comment by breautek** [comment]
> After reviewing the plugin xml file, I realized that it's referencing a build artifact. The typescript files should be getting compiled into www/insets.js (not to be confused with src/www directory).

ATM, I'm assuming the built file is packaged into the npm release, which is why it's not in the github repo.

The `plugin.xml` file is templated. Because I constantly forget to keep the `plugin.xml` version in sync with the package.json file. So it's built during packaging time.

it can be generated by runnng `npm pack` or `npm run prepack`

The JS code can be built by running `npm run build`

File: README.md
Comment (TylerBreau) [reply README.md:17]: With the changes coming in this PR, it's probably better to expect people to platform guard as iOS will receive an useless Insets object and inline style could override their actual css env vars usage.

File: docs.md
Comment (TylerBreau) [reply docs.md]:
```
    public async free(): Promise<void> {
        return await Inset.free(this);
    }
```
There's 2 free APIs, a non-static and a static version. I don't see where the docs note this detail.

File: src/www/IInsetConfiguration.ts
Comment (breautek) [reply src/www/IInsetConfiguration.ts:23]: `>>` is a right shift and `<<` is a left shift.

```
CAPTION_BAR                 = 1,
    DISPLAY_CUTOUT              = 1 << 1,
    IME                         = 1 << 2,
    MANDATORY_SYSTEM_GESTURES   = 1 << 3,
```

the first is equivalent of `1 << 0` aka no shift, so the value is simply 1, or `0b00000001`.

`1 << 1` shifts the bits by 1 from 1, which is `0b00000010`, or 2.
`1 << 2` shifts the bits by 2 from 1, which is `0b00000100`, or 4.

The bit literal shows how the `1` is shifting to the left and how the values correspond.

You could do the math 1, 2, 4, 8, 16, ... or you could just shift the proper spot which is easier to read and easy to know what comes next.

File: docs.md
Comment (TylerBreau) [inline docs.md]: `[free()](#32---free)

Random `. Possibly missing the ending quote for code styling?

Comment (breautek) [reply docs.md]: I originally had it as a inline code block, but then changed it to a link to the API. Looks like I just missed the backtick.

File: docs.md
Comment (breautek) [reply docs.md]: it's not important enough to mention. The key takeaway is when it's done being used it should be freed and that is documented.

Documenting every single way to free something adds bloat to the documentation which should be clear and concise. There's not much value noting the convenience method.

A better place for that is a reference document, which we don't have.

File: docs.md
Comment (TylerBreau) [inline docs.md:136]: You can also call the instance API `free` instead.

###### Signature

```typescript
free(): Promise<void>;
```

Comment (TylerBreau) [reply docs.md:136]: ... github comments really like to conflict with md syntax...

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (breautek) [reply src/android/com/totalpave/cordova/inset/Inset.java:281]: this was already addressed.

File: plugin.xml.ejs
Comment (TylerBreau) [inline plugin.xml.ejs:14]: www/insets.js plural naming

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (TylerBreau) [inline src/android/com/totalpave/cordova/inset/Inset.java:241]: With the changes to Inset construction, will this ever have listeners during pluginInitialize?

File: plugin.xml.ejs
Comment (breautek) [reply plugin.xml.ejs:14]: its just the filename

File: src/android/com/totalpave/cordova/inset/Inset.java
Comment (breautek) [reply src/android/com/totalpave/cordova/inset/Inset.java:241]: No, but that's the global listener function so not sure what you're point is.

Comment (TylerBreau) [reply src/android/com/totalpave/cordova/inset/Inset.java:241]: oh right lol my bad

Comment (breautek) [reply src/android/com/totalpave/cordova/inset/Inset.java:241]: That block runs everytime `this.cordova.getActivity().findViewById(android.R.id.content)` has insets applied, either by an external source like an orientation change, or via a `request` API.

**Review by TylerBreau — APPROVED** [review]
Haven't ran this myself but having thoroughly reviewed it, not seeing any more things of note.

**Comment by seamlink-aalves** [comment]
Hi @breautek and @TylerBreau,

Thank you for this PR. 

I was able to implement the new version. It does in fact add much greater flexibility.

Excelent work.

**Pull request merged**
